db/src/migration.rs

```diff
@@ -0,0 +1,75 @@
+use crate::{db::VERSION_KEY, internal_error, Result};
+use ckb_logger::info;
+use rocksdb::{
+    ops::{Get, Put},
+    OptimisticTransactionDB,
+};
+
+#[derive(Default)]
+pub struct Migrations {
+    migrations: Vec<Box<dyn Migration>>,
+}
+
+impl Migrations {
+    pub fn add_migration(&mut self, migration: Box<dyn Migration>) {
+        self.migrations.push(migration);
+    }
+
+    pub fn migrate(&self, db: &OptimisticTransactionDB) -> Result<()> {
+        let db_version = db
+            .get(VERSION_KEY)
+            .map_err(|err| {
+                internal_error(format!("failed to get the version of database: {}", err))
+            })?
+            .map(|version_bytes| unsafe { String::from_utf8_unchecked(version_bytes.to_vec()) });
+
+        match db_version {
+            Some(v) => self
+                .migrations
+                .iter()
+                .filter(|m| m.version() > v.as_str())
+                .try_for_each(|m| {
+                    info!("Migrating database version to {}", m.version());
+                    m.migrate(db)
```

Comment:
The migration script may forget to save the version. It is safer to set the version here. Also it is more consistent, because the `None` branch will write the version.